### PR TITLE
Implement Keys and Values as views in MapField

### PR DESCRIPTION
**Pull request by jskeet**

### CODE REVIEW

File: csharp/src/Google.Protobuf/Collections/MapField.cs [hunk 1]
```diff
@@ -658,5 +662,81 @@ public int CalculateSize()
...
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+
+            public void CopyTo(Array array, int index)
+            {
+                throw new NotImplementedException();
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: was this the intention?

File: csharp/src/Google.Protobuf/Collections/MapField.cs [hunk 2]
```diff
@@ -221,17 +227,15 @@ public bool TryGetValue(TKey key, out TValue value)
             }
         }
 
-        // TODO: Make these views?
-
         /// <summary>
         /// Gets a collection containing the keys in the map.
         /// </summary>
-        public ICollection<TKey> Keys { get { return list.Select(t => t.Key).ToList(); } }
+        public ICollection<TKey> Keys { get { return new MapView<TKey>(this, pair => pair.Key, ContainsKey); } }
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: I think BCL collections actually cache the view once they lazy-initialize it.

See here:
http://referencesource.microsoft.com/#mscorlib/system/collections/hashtable.cs,836

Also below for Values.

File: csharp/src/Google.Protobuf/Collections/MapField.cs (hunk 1 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: Eek, no, not at all. Thanks, will fix.

File: csharp/src/Google.Protobuf/Collections/MapField.cs (hunk 2 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: Mmm. It's certainly not hard to do, but I'm not sure how much benefit there is - at the cost of a couple of extra fields. Once we eventually fix the earlier TODO and don't allocate a map/list until we need one, map fields should be very cheap in terms of memory.

I guess two fields (16 bytes) isn't a terribly heavy price - it depends on how many maps we end up seeing.

Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: The benefit - 
I think always obtaining the same object once you access property like Keys or Values (or property in general) is just what everyone would expect. And breaking natural expectations is almost always a bad idea.  Let's favor API that behaves in the expected way over saving 16 bytes.

Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: Okay - I don't much mind either way, although I don't have that expectation
and of course it's only detectable via reference equality. (If the returned
objects had some other difference, I would be much more in favour.)

Will do that at the same time as the CopyTo fix.

Jon

Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: Sg.
Actually, when thinking of this more in detail - caching the reference doesn't guarantee you will always get the object - if two threads call this and you are unlucky, both will obtain a different reference and only one of them would be cached. So, given it's impossible to guarantee only one reference will ever be returned, the expectation I was living with turns out to be wrong.

Based on this realization, I don't feel strongly about this. Feel free to choose approach that makes most sense to you.

Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: Okay. Will leave it for now, and we can revisit if anyone ever requests it.

**Comment by jskeet** [comment]
Done - PTAL.

File: csharp/src/Google.Protobuf/Collections/MapField.cs [hunk 3]
```diff
@@ -658,5 +662,92 @@ public int CalculateSize()
...
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+
+            public void CopyTo(Array array, int index)
+            {
+                if (index < 0)
+                {
+                    throw new ArgumentOutOfRangeException("arrayIndex");
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: arrayIndex vs index

**Comment by jtattermusch** [comment]
one nit - otherwise LGTM.

File: csharp/src/Google.Protobuf/Collections/MapField.cs (hunk 3 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: Doh - this is why I want to move to C# 6, so I can start using nameof everywhere :)

**Pull request merged**